File: prayer_app_expo/src/services/quranStorageService.js
import AsyncStorage from '@react-native-async-storage/async-storage';
import { pointerKey } from '../models/quranModels';

export const LAST_READ_KEY = 'quran_last_read';
export const RECENTS_KEY = 'quran_recents';
export const BOOKMARKS_KEY = 'quran_bookmarks';

export async function loadLastRead() {
    try {
        const raw = await AsyncStorage.getItem(LAST_READ_KEY);
        return raw ? JSON.parse(raw) : null;
    } catch (_) {
        return null;
    }
}

export async function setLastRead(pointer) {
    await AsyncStorage.setItem(LAST_READ_KEY, JSON.stringify(pointer));
}

export async function loadRecents() {
    try {
        const raw = await AsyncStorage.getItem(RECENTS_KEY);
        return raw ? JSON.parse(raw) : [];
    } catch (_) {
        return [];
    }
}

export async function pushRecent(pointer) {
    const recents = await loadRecents();
    const key = pointerKey(pointer);
    const deduped = recents.filter((r) => pointerKey(r) !== key);
    deduped.unshift(pointer);
    const sliced = deduped.slice(0, 10);
    await AsyncStorage.setItem(RECENTS_KEY, JSON.stringify(sliced));
    return sliced;
}

export async function loadBookmarks() {
    try {
        const raw = await AsyncStorage.getItem(BOOKMARKS_KEY);
        return raw ? JSON.parse(raw) : [];
    } catch (_) {
        return [];
    }
}

export async function toggleBookmark(pointer) {
    const bookmarks = await loadBookmarks();
    const key = pointerKey(pointer);
    const idx = bookmarks.findIndex((b) => pointerKey(b) === key);
    if (idx >= 0) {
        bookmarks.splice(idx, 1);
    } else {
        bookmarks.unshift(pointer);
    }
    await AsyncStorage.setItem(BOOKMARKS_KEY, JSON.stringify(bookmarks));
    return bookmarks;
}

export async function removeBookmark(pointer) {
    const bookmarks = await loadBookmarks();
    const key = pointerKey(pointer);
    const next = bookmarks.filter((b) => pointerKey(b) !== key);
    await AsyncStorage.setItem(BOOKMARKS_KEY, JSON.stringify(next));
    return next;
}

export function isBookmarked(bookmarks, pointer) {
    return (bookmarks || []).some((b) => pointerKey(b) === pointerKey(pointer));
}

File: prayer_app_flutter/lib/src/services/quran_storage_service.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';
import '../models/quran_models.dart';

class QuranStorageService {
  static const String lastReadKey = 'quran_last_read';
  static const String recentsKey = 'quran_recents';
  static const String bookmarksKey = 'quran_bookmarks';

  Future<QuranPointer?> loadLastRead() async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString(lastReadKey);
    if (raw == null) return null;
    try {
      return QuranPointer.fromJson(jsonDecode(raw) as Map<String, dynamic>);
    } catch (_) {
      return null;
    }
  }

  Future<List<QuranPointer>> loadRecents() async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString(recentsKey);
    if (raw == null) return [];
    try {
      final list = jsonDecode(raw) as List<dynamic>;
      return list
          .whereType<Map<String, dynamic>>()
          .map(QuranPointer.fromJson)
          .toList();
    } catch (_) {
      return [];
    }
  }

  Future<List<QuranPointer>> loadBookmarks() async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString(bookmarksKey);
    if (raw == null) return [];
    try {
      final list = jsonDecode(raw) as List<dynamic>;
      return list
          .whereType<Map<String, dynamic>>()
          .map(QuranPointer.fromJson)
          .toList();
    } catch (_) {
      return [];
    }
  }

  Future<void> setLastRead(QuranPointer pointer) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(lastReadKey, jsonEncode(pointer.toJson()));
  }

  Future<List<QuranPointer>> pushRecent(QuranPointer pointer) async {
    final recents = await loadRecents();
    recents.removeWhere((r) => r.dedupeKey() == pointer.dedupeKey());
    recents.insert(0, pointer);
    if (recents.length > 10) {
      recents.removeRange(10, recents.length);
    }
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(recentsKey, jsonEncode(recents.map((e) => e.toJson()).toList()));
    return recents;
  }

  Future<List<QuranPointer>> toggleBookmark(QuranPointer pointer) async {
    final bookmarks = await loadBookmarks();
    final index = bookmarks.indexWhere((b) => b.dedupeKey() == pointer.dedupeKey());
    if (index >= 0) {
      bookmarks.removeAt(index);
    } else {
      bookmarks.insert(0, pointer);
    }
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(bookmarksKey, jsonEncode(bookmarks.map((e) => e.toJson()).toList()));
    return bookmarks;
  }

  Future<List<QuranPointer>> removeBookmark(QuranPointer pointer) async {
    final bookmarks = await loadBookmarks();
    bookmarks.removeWhere((b) => b.dedupeKey() == pointer.dedupeKey());
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(bookmarksKey, jsonEncode(bookmarks.map((e) => e.toJson()).toList()));
    return bookmarks;
  }

  bool isBookmarked(List<QuranPointer> bookmarks, QuranPointer pointer) {
    return bookmarks.any((b) => b.dedupeKey() == pointer.dedupeKey());
  }
}
